### PR TITLE
docs: improve jsdocs

### LINKED ISSUE #412: Improve JSDocs
### Feature

By switching GCommands to TypeScript we stopped using JSDocs which is a mistake. With JSDocs you can interact even more with the user and make the user experience even better.

### Ideal solution or implementation

One by one, we will document each file with JSDocs. We are working on adding JSDocs as fast as possible.

### Alternative solutions or implementations

- [ ] AutocompleteHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/handlers/AutocompleteHandler.ts)
- [ ] ComponentHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/handlers/ComponentHandler.ts)
- [ ] CooldownHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/handlers/CooldownHandler.ts)
- [ ] InteractionCommandHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/handlers/InteractionCommandHandler.ts)
- [ ] MessageCommandHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/handlers/MessageCommandHandler.ts)
- [ ] index.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/index.ts)
- [ ] ChannelOnly.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/ChannelOnly.ts)
- [ ] ClientPermissions.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/ClientPermissions.ts)
- [ ] ClientRoles.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/ClientRoles.ts)
- [ ] Confirm.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/Confirm.ts)
- [ ] index.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/index.ts)
- [ ] Inhibitor.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/Inhibitor.ts)
- [ ] MemberPermissions.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/MemberPermissions.ts)
- [ ] MemberRoles.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/MemberRoles.ts)
- [ ] Nsfw.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/Nsfw.ts)
- [ ] Or.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/Or.ts)
- [ ] UserOnly.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/inhibitors/UserOnly.ts)
- [x] GClient.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/GClient.ts)
   - https://github.com/Garlic-Team/gcommands/pull/413
- [x] directoryLoader.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/loaders/directoryLoader.ts)
- [x] pluginFinder.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/loaders/pluginFinder.ts)
- [ ] CommandManager.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/managers/CommandManager.ts)
- [ ] ComponentManager.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/managers/ComponentManager.ts)
- [ ] HandlerManager.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/managers/HandlerManager.ts)
- [ ] ListenerManager.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/managers/ListenerManager.ts)
- [ ] PluginManager.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/managers/PluginManager.ts)
- [ ] Argument.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Argument.ts)
- [ ] Attachment.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Attachment.ts)
- [ ] base.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/base.ts)
- [ ] Boolean.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Boolean.ts)
- [ ] Channel.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Channel.ts)
- [ ] Integer.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Integer.ts)
- [ ] Mentionable.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Mentionable.ts)
- [ ] Number.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Number.ts)
- [ ] Role.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/Role.ts)
- [ ] String.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/String.ts)
- [ ] User.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/arguments/User.ts)
- [ ] Command.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Command.ts)
- [ ] Component.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Component.ts)
- [ ] AutocompleteContext.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/contexts/AutocompleteContext.ts)
- [ ] CommandContext.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/contexts/CommandContext.ts)
- [ ] ComponentContext.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/contexts/ComponentContext.ts)
- [ ] Context.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/contexts/Context.ts)
- [ ] Listener.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Listener.ts)
- [ ] Plugin.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Plugin.ts)
- [ ] Provider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/structures/Provider.ts)
- [ ] common.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/common.ts)
- [ ] Logger.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/logger/Logger.ts)
- [x] regexes.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/regexes.ts)
   - https://github.com/Garlic-Team/gcommands/pull/411
- [ ] registerDirectories.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/registerDirectories.ts)
- [ ] registerDirectory.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/registerDirectory.ts)
- [ ] sync.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/sync.ts)
- [x] Util.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/lib/util/Util.ts)
   - https://github.com/Garlic-Team/gcommands/pull/413 
- [ ] AutocompleteHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/listeners/AutocompleteHandler.ts)
- [ ] ComponentHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/listeners/ComponentHandler.ts)
- [ ] InteractionCommandHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/listeners/InteractionCommandHandler.ts)
- [ ] MessageCommandHandler.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/listeners/MessageCommandHandler.ts)
- [ ] Ready.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/listeners/Ready.ts)
- [ ] FirestoreProvider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/providers/FirestoreProvider.ts)
- [ ] KeyvProvider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/providers/KeyvProvider.ts)
- [ ] LruCacheProvider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/providers/LruCacheProvider.ts)
- [ ] MongoDBProvider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/providers/MongoDBProvider.ts)
- [ ] PrismaIOProvider.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/providers/PrismaIOProvider.ts)
- [ ] confirm.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/util/confirm.ts)
- [ ] customId.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/util/customId.ts)
- [ ] uid.ts (https://github.com/Garlic-Team/gcommands/blob/next/src/util/uid.ts)

### Other context

_No response_

### PR DESCRIPTION
**Pull request by xhyrom**
**Please describe the changes this PR makes and why it should be merged:**
https://github.com/Garlic-Team/gcommands/issues/412
- [ ] ESlint JSDocs rule

**Status and versioning classification:**
- This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Review by S222em — COMMENTED** [review]
Should this not include jsdoc for all files?

**Comment by xhyrom** [comment]
> Should this not include jsdoc for all files?

This will include, it's draft

File: src/lib/GClient.ts [hunk 1]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 */
 	messageSupport?: boolean;
+	/**
+	 * Prefix for message commands
+	 * @requires {@link GClientOptions.messageSupport} to be enabled
+	 * @type {string}
+	 */
 	messagePrefix?: string;
+	/**
+	 * Whether to send a message for a unknown command.
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * Whether to send a message for an unknown command.
```

File: src/lib/GClient.ts [hunk 2]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 * @type {boolean}
+	 */
 	unknownCommandMessage?: boolean;
+	/**
+	 * Array of all folders to be loaded by GCommands.
+	 *
+	 * It can load from these folders:
+	 * * Commands
+	 * * Listeners (Events)
+	 * * Components
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * GCommands will check all the files in each folder for these classes:
	 * * Command
	 * * Listener
	 * * Component
```

File: src/lib/GClient.ts [hunk 3]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 *
+	 * @type {string[]}
+	 * @example
+	 * ```typescript
+	 * new GClient({
+	 * 		dirs: [
+	 * 			'commands',
+	 * 			'events',
+	 * 			'components',
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * 			join(__dirname, 'commands'),
	 * 			join(__dirname, 'events'),
	 * 			join(__dirname, 'components'),
```

File: src/lib/GClient.ts [hunk 4]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 * 		]
+	 * })
+	 * ```
+	 */
 	dirs?: Array<string>;
+	/**
+	 * The database to be used in the project.
+	 *
+	 * You can put whatever you want in this option, it only serves your purposes.
+	 * The library itself does not use the database.
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * This option is only for easier access and is not used by GCommands.
```

File: src/lib/GClient.ts [hunk 5]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 *
+	 * @type {any}
+	 */
 	database?: any;
-	devGuildId?: string;
+	/**
+	 * The guild id that will serve as development server.
+	 *
+	 * All application commands are not automatically loaded as global, but only as guildOnly for this option.
+	 * If you want to make the commands public, you need to delete this options.
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * The guild id that will serve as the development server for your application.
	 *
	 * If this option is present, all slash commands will be registered only for the specified guild id.
```

File: src/lib/GClient.ts [hunk 6]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+	 */
+	devGuildId?: Snowflake;
 }
 
+/**
+ * The base {@link Client} extension that makes GCommands work. To use GCommands, you must use this class or extend it.
+ *
+ * GCommands automatically loads everything.
+ *
+ * @see {@link GClientOptions} for all available options in GClient.
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
 * The base {@link Client} that GCommands uses.
 *
 * @see {@link GClientOptions} for all available options for GClient.
```

File: src/lib/GClient.ts [hunk 7]
```diff
@@ -22,17 +29,90 @@ export enum AutoDeferType {
...
+ *
+ * @extends {Client}
+ */
 export class GClient<Ready extends boolean = boolean> extends Client<Ready> {
+	/**
+	 * Object of all basic messages if a problem occurs.
+	 *
+	 * However, you can customize these messages using the [@gcommands/plugin-language](https://github.com/Garlic-Team/gcommands-addons/tree/master/packages/plugin-language) plugin.
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	 * Object of the default responses that GCommands uses for auto responding in the case of something happening.
	 *
	 * You can customize these messages using the [@gcommands/plugin-language](https://github.com/Garlic-Team/gcommands-addons/tree/master/packages/plugin-language) plugin.
```

File: src/lib/GClient.ts [hunk 8]
```diff
@@ -1,12 +1,19 @@
...
 import { Commands } from './managers/CommandManager';
 import { Components } from './managers/ComponentManager';
 import { Listeners } from './managers/ListenerManager';
 import { Plugins } from './managers/PluginManager';
 import { registerDirectories } from './util/registerDirectories';
 import Responses from '../responses.json';
 
+/**
+ * Enum for auto defer feature.
+ * Automatic defer if bot does not reply within more than 3s
```
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
 * Enum for the auto defer feature.
 *
 * Automatically defers interaction if application doesn't respond in 3s
```

File: src/lib/GClient.ts [hunk 9]
```diff
@@ -55,6 +136,35 @@ export class GClient<Ready extends boolean = boolean> extends Client<Ready> {
...
+	 * const client = new GClient({
+	 * 		..settings,
+	 * 		database: new MongoDBProvider('mongodb://localhost:27017/database')
+	 * })
+	 *
+	 * const db = client.getDatabase(MongoDBProvider.prototype);
+	 * // Types working in `db`
+	 * ```
+	 */
 	public getDatabase<Database>(_?: Database): Database {
```
Comment (Gabe616) [inline src/lib/GClient.ts:167]:
```suggestion
	 * const db = client.getDatabase();
	 * // returns <MongoDBProvider>
	 * ```
	 * @example JavaScript example
	 * ```javascript
	 * const { MongoDBProvider } = require('gcommands/dist/providers/MongoDBProvider');
	 *
	 * const client = new GClient({
	 * 		..settings,
	 * 		database: new MongoDBProvider('mongodb://localhost:27017/database')
	 * })
	 *
	 * const db = client.getDatabase();
	 * // returns <MongoDBProvider>
	 * ```
	 */
	public getDatabase<Database>(): Database {
```

**Review by nexpid — CHANGES_REQUESTED** [review]
fixed some stuff yay

File: src/lib/GClient.ts (hunk 9 above)
Comment (nexpid) [inline src/lib/GClient.ts]:
```suggestion
	/**
	 * The method that returns the database option provided in {@link GClientOptions}
	 * @param {any} _ Used for typings
	 * @returns {Database}
	 * @example TypeScript example
	 * ```typescript
	 * import { MongoDBProvider } from 'gcommands/dist/providers/MongoDBProvider';
	 *
	 * const client = new GClient({
	 * 		..settings,
	 * 		database: new MongoDBProvider('mongodb://localhost:27017/database')
	 * })
	 *
	 * const db = client.getDatabase(MongoDBProvider.prototype);
	 * // returns <MongoDBProvider>
	 * ```
	 * @example JavaScript example
	 * ```javascript
	 * const { MongoDBProvider } = require('gcommands/dist/providers/MongoDBProvider');
	 *
	 * const client = new GClient({
	 * 		..settings,
	 * 		database: new MongoDBProvider('mongodb://localhost:27017/database')
	 * })
	 *
	 * const db = client.getDatabase(MongoDBProvider.prototype);
	 * // returns <MongoDBProvider>
	 * ```
	 */
	public getDatabase<Database>(_?: Database): Database {
```

**Review by nexpid — CHANGES_REQUESTED** [review]
updated getDatabase

File: src/lib/GClient.ts [hunk 10]
```diff
@@ -22,17 +30,88 @@ export enum AutoDeferType {
...
+	 */
 	messageSupport?: boolean;
+	/**
+	 * Prefix for message commands
+	 * @requires {@link GClientOptions.messageSupport} to be enabled
+	 * @type {string}
+	 */
 	messagePrefix?: string;
+	/**
+	 * Whether to send a message for an unknown command.
```
Comment (S222em) [inline src/lib/GClient.ts]: This should state that it only works for message commands. 
```suggestion
	 * Whether to send a message for an unknown message command.
```

**Comment by xhyrom** [comment]
Please review @S222em . I'll finish the others in time, but not to delay v9.2.0 unnecessarily and at least have JSDocs on some properties.

File: src/index.ts [hunk 11]
```diff
@@ -33,8 +33,9 @@ export * from './lib/util/logger/Logger';
 // Inhibitors
 export * as Inhibitor from './inhibitors';
 
-/* Providers
- * Providers will not be exported due to additional modules. To import, just use the path gcommands/dist/providers/{name}
+/**
+ * Providers will not be exported due to additional modules.
```
Comment (S222em) [inline src/index.ts]:
```suggestion
 * Providers are not exported
```

Dont think this really needs a reason

File: src/lib/GClient.ts [hunk 12]
```diff
@@ -24,19 +44,90 @@ export enum AutoDeferType {
...
+	 * 			join(__dirname, 'commands'),
+	 * 			join(__dirname, 'events'),
+	 * 			join(__dirname, 'components'),
+	 * 		]
+	 * })
+	 * ```
+	 */
 	dirs?: Array<string>;
+	/**
+	 * The database to be used in the project.
```
Comment (S222em) [inline src/lib/GClient.ts:91]:
```suggestion
	 * Database for this client
```

File: src/lib/GClient.ts [hunk 13]
```diff
@@ -24,19 +44,90 @@ export enum AutoDeferType {
...
+	 * This option is only for easier access and is not used by GCommands.
+	 *
+	 * @type {any}
+	 */
 	database?: any;
-	devGuildId?: string;
+	/**
+	 * The guild id that will serve as the development server for your application.
+	 *
+	 * If this option is present, all slash commands will be registered only for the specified guild id.
```
Comment (S222em) [inline src/lib/GClient.ts:101]: guildId set in the command will override it

**Review by S222em — CHANGES_REQUESTED** [review]

File: src/lib/GClient.ts (hunk 12 above)
Comment (xhyrom) [reply src/lib/GClient.ts:91]: It's not for client because gcommands doesn't using db

**Pull request merged**